dask/array/routines.py

```diff
@@ -531,10 +533,14 @@ def round(a, decimals=0):
 
 @wraps(np.unique)
 def unique(x):
-    name = 'unique-' + x.name
-    dsk = dict(((name, i), (np.unique, key)) for i, key in enumerate(x._keys()))
-    parts = Array._get(sharedict.merge((name, dsk), x.dask), list(dsk.keys()))
-    return np.unique(np.concatenate(parts))
+    x = x.ravel()
+
+    out = atop(np.unique, "i", x, "i", dtype=x.dtype)
+    out._chunks = tuple((np.nan,) * len(c) for c in out.chunks)
+
+    out = from_delayed(delayed(np.unique)(out), (np.nan,), x.dtype)
```

Comment:
As it's illegal to rechunk when we don't know the chunking/size, this was the easiest thing I could think of as a workaround. It basically reproduces the older behavior, but in a lazy way. Though if there are better suggestions, would be open to hearing them.

Reply:
Rather than use delayed (which will do things like trigger optimize calls) we might instead consider adding the following task to the graph:

```python
dsk = {key: (np.unique, (np.concatenate, out._keys().tolist()))}
```

Reply:
If we cared about large arrays then we might consider a tree reduction here (there might already be nice code for this in `reductions.py`?)

Reply:
Thanks. Was thinking there should be a Dask graph-style solution, but was unsure what it was. Will give that a try.

At least for me, I don't think tree reductions really matter here. Given this implementation has sat for a long time the same way, I'm guessing not that many other people care either. That said, I'd be willing take a look. Do you have some more specific examples?

Reply:
Yeah, I didn't mean to imply that this should happen before this is merged.  Just something to think about in the future.  

I think that most (all?) of the reduction code in any of array/bag/dataframes do tree-reductions these days.

Reply:
Oh ... 😆 ... I think I got the graph solution to work anyways. Care to take a look? 😉

Reply:
Sorry, I did mean for the graph thing to go in (switching collection types is not free).  I meant to say that the tree-reduction solution probably wasn't necessary near-term.

Reply:
Ah ok. Agreed.

Reply:
As for tree reductions with `unique`, have raised issue ( https://github.com/dask/dask/issues/2851 ) to track this desired change.